### PR TITLE
Change Version formatting properties and functions to return Version objects

### PR DESCRIPTION
**Pull request by adamjstewart**
Also added better API docs.

In #4823, @akthoma discovered an interesting URL:
https://www.tbi.univie.ac.at/RNA/download/sourcecode/2_3_x/ViennaRNA-2.3.5.tar.gz

In order to properly compute this URL based on the version number, we need to be able to do the following:
```python
def url_for_version(self, version):
    url = "https://www.tbi.univie.ac.at/RNA/download/sourcecode/2_3_x/ViennaRNA-2.3.5.tar.gz"
    return url.format(version.up_to(2).underscored, version.dotted)
```
Previously, this wasn't possible because `version.up_to()` returned a string. Now that it returns a `Version` object, it is possible to string together `up_to()` with `dotted`, `dashed`, `underscored`, and `joined`.

### CODE REVIEW

**Comment by adamjstewart** [comment]
I didn't test any packages yet. Let me take a look.

**Comment by adamjstewart** [comment]
Okay, that should be all of the packages that needed to be fixed. The rest all use `'{0}'.format(version.up_to())` in some form or another.

Also added unit tests for `up_to()`. Apparently we didn't have any before.

**Comment by adamjstewart** [comment]
There's one other thing bugging me about our version formatting support. This was the case before this PR, so it doesn't necessarily need to be tackled in this PR.
```python
>>> version = Version('1.2.3b')
>>> version.dotted
'1.2.3.b'  # I would expect '1.2.3b' here
>>> version.dashed
'1-2-3-b'  # I would expect '1-2-3b' here
>>> version.underscored
'1_2_3_b'  # I would expect '1_2_3b' here
```
Also:
```python
>>> version = Version('1.2b.3')
>>> version.up_to(2)
Version('1.2')      # I would expect Version('1.2b') here, but this isn't that surprising
>>> version.up_to(3)
Version('1.2.b')    # I would expect Version('1.2b') or Version('1.2b.3') here
>>> version.up_to(4)
Version('1.2.b.3')  # I would expect Version('1.2b.3') here
```
In the cases of `dotted` and `up_to(4)`, what you get out isn't what you put in even though they should be identical. This worries me. It's amazing this hasn't come up yet.

I understand that we need to separate these out at character/digit boundaries for comparison purposes, but maybe we should keep them together for string formatting?

**Comment by tgamblin** [comment]
@adamjstewart: the original separators are stored in the `separators` property, and the original string is preserve in the `string` property.  You *could* make `dotted` and friends only insert a separator if it was present in the original string.

**Comment by adamjstewart** [comment]
Oh yeah, that's another thing that is weird:
```python
>>> version = Version('1_2_3')
>>> version.up_to(2)
Version('1.2')
```
We shouldn't really be changing the separator character here.

I'll think about it, but I don't want this to hold up the PR. We can discuss this in another PR if you're okay with the current changes.

**Comment by adamjstewart** [comment]
A more complex case that needs to work is versions like `1.2-3.4`, which is common in R packages. `version.up_to(3)` needs to be `Version('1.2-3')`.

File: lib/spack/spack/version.py [hunk 1]
```diff
@@ -204,11 +262,9 @@ def __getitem__(self, idx):
             return self.version[idx]
 
         elif isinstance(idx, slice):
-            # Currently len(self.separators) == len(self.version) - 1
-            extendend_separators = self.separators + ('',)
             string_arg = []
 
-            pairs = zip(self.version[idx], extendend_separators[idx])
+            pairs = zip(self.version[idx], self.separators[idx])
```
Comment (adamjstewart) [inline lib/spack/spack/version.py]: @tgamblin This seems to be the only place in Spack that uses `version.separators`. I changed the implementation above to make this simpler. Let me know if you want me to undo this change.

**Comment by adamjstewart** [comment]
Okay, I fixed all of the weird corner cases I mentioned above. I didn't bother changing whether `Version('1.2b.3').up_to(2)` returns `Version('1.2')` or `Version('1.2b')` as I couldn't think of any use cases where this is actually required.

Now the real question is whether or not `dotted`, `dashed`, `underscored`, and `joined` return `str` or `Version`. If we chose the latter, both `version.up_to(2).underscored` and `version.underscored.up_to(2)` would work the same.

**Comment by tgamblin** [comment]
I think making them return `Version` would be good to avoid confusion.

**Comment by adamjstewart** [comment]
Just remembered that:
```python
>>> Version('1.2.3') == Version('1-2-3') == Version('1_2_3')
True
```
so my previous tests weren't actually that useful. Updated them to test the string value instead.
